File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/TranslateVisitor.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tinkerpop.gremlin.language.translator;

import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;
import org.apache.tinkerpop.gremlin.language.grammar.GremlinBaseVisitor;
import org.apache.tinkerpop.gremlin.language.grammar.GremlinParser;
import org.apache.tinkerpop.gremlin.process.traversal.DT;
import org.apache.tinkerpop.gremlin.process.traversal.Merge;
import org.apache.tinkerpop.gremlin.process.traversal.Operator;
import org.apache.tinkerpop.gremlin.process.traversal.Order;
import org.apache.tinkerpop.gremlin.process.traversal.P;
import org.apache.tinkerpop.gremlin.process.traversal.Pick;
import org.apache.tinkerpop.gremlin.process.traversal.Pop;
import org.apache.tinkerpop.gremlin.process.traversal.SackFunctions;
import org.apache.tinkerpop.gremlin.process.traversal.Scope;
import org.apache.tinkerpop.gremlin.process.traversal.TextP;
import org.apache.tinkerpop.gremlin.structure.Column;
import org.apache.tinkerpop.gremlin.structure.Direction;
import org.apache.tinkerpop.gremlin.structure.T;
import org.apache.tinkerpop.gremlin.structure.VertexProperty.Cardinality;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * A Gremlin to Gremlin translator. Makes no changes to input except:
 * <ul>
 *     <li>Normalizes whitespace</li>
 *     <li>Normalize numeric suffixes to lower case</li>
 *     <li>Makes anonymous traversals explicit with double underscore</li>
 *     <li>Makes enums explicit with their proper name</li>
 * </ul>
 */
public class TranslateVisitor extends GremlinBaseVisitor<Void> {

    protected final String graphTraversalSourceName;

    protected final StringBuilder sb = new StringBuilder();

    protected final Set<String> parameters = new HashSet<>();

    public TranslateVisitor() {
        this("g");
    }

    public TranslateVisitor(final String graphTraversalSourceName) {
        this.graphTraversalSourceName = graphTraversalSourceName;
    }

    public String getTranslated() {
        return sb.toString();
    }

    public Set<String> getParameters() {
        return parameters;
    }

    protected String processGremlinSymbol(final String step) {
        return step;
    }

    protected void appendArgumentSeparator() {
        sb.append(", ");
    }

    protected void appendStepSeparator() {
        sb.append(".");
    }

    protected void appendStepOpen() {
        sb.append("(");
    }

    protected void appendStepClose() {
        sb.append(")");
    }

    protected static String removeFirstAndLastCharacters(final String text) {
        return text != null && !text.isEmpty() ? text.substring(1, text.length() - 1) : "";
    }

    @Override
    public Void visitTraversalSource(final GremlinParser.TraversalSourceContext ctx) {
        // replace "g" with whatever the user wanted to use as the graph traversal source name
        sb.append(graphTraversalSourceName);

        // child counts more than 1 means there is a step separator and a traversal source method
        if (ctx.getChildCount() > 1) {
            if (ctx.getChild(0).getChildCount() > 1) {
                appendStepSeparator();
                visitTraversalSourceSelfMethod((GremlinParser.TraversalSourceSelfMethodContext) ctx.getChild(0).getChild(2));
            }
            appendStepSeparator();
            visitTraversalSourceSelfMethod((GremlinParser.TraversalSourceSelfMethodContext) ctx.getChild(2));
        }
        return null;
    }

    @Override
    public Void visitNestedTraversal(final GremlinParser.NestedTraversalContext ctx) {
        if (ctx.ANON_TRAVERSAL_ROOT() == null)
            appendAnonymousSpawn();
        return visitChildren(ctx);
    }

    @Override
    public Void visitTraversalScope(final GremlinParser.TraversalScopeContext ctx) {
        appendExplicitNaming(ctx.getText(), Scope.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalT(final GremlinParser.TraversalTContext ctx) {
        appendExplicitNaming(ctx.getText(), T.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalMerge(final GremlinParser.TraversalMergeContext ctx) {
        appendExplicitNaming(ctx.getText(), Merge.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalOrder(final GremlinParser.TraversalOrderContext ctx) {
        appendExplicitNaming(ctx.getText(), Order.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalBarrier(final GremlinParser.TraversalBarrierContext ctx) {
        appendExplicitNaming(ctx.getText(), SackFunctions.Barrier.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalDirection(final GremlinParser.TraversalDirectionContext ctx) {
        appendExplicitNaming(ctx.getText(), Direction.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalCardinality(final GremlinParser.TraversalCardinalityContext ctx) {
        // handle the enum style of cardinality if there are no parens otherwise is the function call.
        if (ctx.LPAREN() != null && ctx.RPAREN() != null) {
            final int idx = ctx.K_CARDINALITY() != null ? 2 : 0;
            final String txt = ctx.getChild(idx).getText();
            appendExplicitNaming(txt, getCardinalityFunctionClass());
            appendStepOpen();
            visit(ctx.genericLiteral());
            appendStepClose();
        } else {
            appendExplicitNaming(ctx.getText(), "Cardinality");
        }

        return null;
    }

    @Override
    public Void visitTraversalColumn(final GremlinParser.TraversalColumnContext ctx) {
        appendExplicitNaming(ctx.getText(), Column.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalPop(final GremlinParser.TraversalPopContext ctx) {
        appendExplicitNaming(ctx.getText(), Pop.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalOperator(final GremlinParser.TraversalOperatorContext ctx) {
        appendExplicitNaming(ctx.getText(), Operator.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalPick(final GremlinParser.TraversalPickContext ctx) {
        appendExplicitNaming(ctx.getText(), Pick.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalDT(final GremlinParser.TraversalDTContext ctx) {
        appendExplicitNaming(ctx.getText(), DT.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalPredicate(final GremlinParser.TraversalPredicateContext ctx) {
        switch(ctx.getChildCount()) {
            case 1:
                // handle simple predicate
                visit(ctx.getChild(0));
                break;
            case 5:
                // handle negate of P
                visit(ctx.getChild(0));
                sb.append(".").append(processGremlinSymbol("negate")).append("()");
                break;
            case 6:
                // handle and/or predicates
                final int childIndexOfParameterOperator = 2;
                final int childIndexOfCaller = 0;
                final int childIndexOfArgument = 4;

                visit(ctx.getChild(childIndexOfCaller));
                sb.append(".").append(processGremlinSymbol(ctx.getChild(childIndexOfParameterOperator).getText())).append("(");
                visit(ctx.getChild(childIndexOfArgument));
                sb.append(")");
                break;
        }
        return null;
    }

    @Override
    public Void visitTraversalPredicate_eq(final GremlinParser.TraversalPredicate_eqContext ctx) {
        visitP(ctx, P.class, "eq");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_neq(final GremlinParser.TraversalPredicate_neqContext ctx) {
        visitP(ctx, P.class, "neq");
        return null;
    }

    protected void visitP(final ParserRuleContext ctx, final Class<?> clazzOfP, final String methodName) {
        sb.append(clazzOfP.getSimpleName());
        appendStepSeparator();
        sb.append(processGremlinSymbol(methodName));
        appendStepOpen();

        final List<ParseTree> list =  ctx.children.stream().filter(
                t -> t instanceof GremlinParser.GenericLiteralArgumentContext ||
                              t instanceof GremlinParser.GenericLiteralListArgumentContext ||
                              t instanceof GremlinParser.StringArgumentContext ||
                              t instanceof GremlinParser.TraversalPredicateContext).collect(Collectors.toList());
        for (int ix = 0; ix < list.size(); ix++) {
            visit(list.get(ix));
            if (ix < list.size() - 1) appendArgumentSeparator();
        }

        appendStepClose();
    }

    @Override
    public Void visitTraversalPredicate_lt(final GremlinParser.TraversalPredicate_ltContext ctx) {
        visitP(ctx, P.class, "lt");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_lte(final GremlinParser.TraversalPredicate_lteContext ctx) {
        visitP(ctx, P.class, "lte");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_gt(final GremlinParser.TraversalPredicate_gtContext ctx) {
        visitP(ctx, P.class, "gt");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_gte(final GremlinParser.TraversalPredicate_gteContext ctx) {
        visitP(ctx, P.class, "gte");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_inside(final GremlinParser.TraversalPredicate_insideContext ctx) {
        visitP(ctx, P.class, "inside");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_outside(final GremlinParser.TraversalPredicate_outsideContext ctx) {
        visitP(ctx, P.class, "outside");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_between(final GremlinParser.TraversalPredicate_betweenContext ctx) {
        visitP(ctx, P.class, "between");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_within(final GremlinParser.TraversalPredicate_withinContext ctx) {
        visitP(ctx, P.class, "within");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_without(final GremlinParser.TraversalPredicate_withoutContext ctx) {
        visitP(ctx, P.class, "without");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_not(final GremlinParser.TraversalPredicate_notContext ctx) {
        visitP(ctx, P.class, "not");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_containing(final GremlinParser.TraversalPredicate_containingContext ctx) {
        visitP(ctx, TextP.class, "containing");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_notContaining(final GremlinParser.TraversalPredicate_notContainingContext ctx) {
        visitP(ctx, TextP.class, "notContaining");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_startingWith(final GremlinParser.TraversalPredicate_startingWithContext ctx) {
        visitP(ctx, TextP.class, "startingWith");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_notStartingWith(final GremlinParser.TraversalPredicate_notStartingWithContext ctx) {
        visitP(ctx, TextP.class, "notStartingWith");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_endingWith(final GremlinParser.TraversalPredicate_endingWithContext ctx) {
        visitP(ctx, TextP.class, "endingWith");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_notEndingWith(final GremlinParser.TraversalPredicate_notEndingWithContext ctx) {
        visitP(ctx, TextP.class, "notEndingWith");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_regex(final GremlinParser.TraversalPredicate_regexContext ctx) {
        visitP(ctx, TextP.class, "regex");
        return null;
    }

    @Override
    public Void visitTraversalPredicate_notRegex(final GremlinParser.TraversalPredicate_notRegexContext ctx) {
        visitP(ctx, TextP.class, "notRegex");
        return null;
    }

    @Override
    public Void visitBooleanArgument(final GremlinParser.BooleanArgumentContext ctx) {
        if (ctx.booleanLiteral() != null)
            visitBooleanLiteral(ctx.booleanLiteral());
        else
            visitVariable(ctx.variable());

        return null;
    }

    @Override
    public Void visitGenericLiteralArgument(final GremlinParser.GenericLiteralArgumentContext ctx) {
        if (ctx.genericLiteral() != null)
            visitGenericLiteral(ctx.genericLiteral());
        else
            visitVariable(ctx.variable());

        return null;
    }

    @Override
    public Void visitIntegerLiteral(final GremlinParser.IntegerLiteralContext ctx) {
        sb.append(ctx.getText().toLowerCase());
        return null;
    }

    @Override
    public Void visitFloatLiteral(final GremlinParser.FloatLiteralContext ctx) {
        if (ctx.infLiteral() != null) return visit(ctx.infLiteral());
        if (ctx.nanLiteral() != null) return visit(ctx.nanLiteral());
        sb.append(ctx.getText().toLowerCase());
        return null;
    }

    @Override
    public Void visitBooleanLiteral(final GremlinParser.BooleanLiteralContext ctx) {
        sb.append(ctx.getText());
        return null;
    }

    @Override
    public Void visitNullLiteral(final GremlinParser.NullLiteralContext ctx) {
        sb.append(ctx.getText());
        return null;
    }

    @Override
    public Void visitNanLiteral(final GremlinParser.NanLiteralContext ctx) {
        sb.append(ctx.getText());
        return null;
    }

    @Override
    public Void visitInfLiteral(final GremlinParser.InfLiteralContext ctx) {
        sb.append(ctx.getText());
        return null;
    }

    @Override
    public Void visitUuidLiteral(final GremlinParser.UuidLiteralContext ctx) {
        sb.append(ctx.getText());
        return null;
    }

    @Override
    public Void visitVariable(final GremlinParser.VariableContext ctx) {
        final String var = ctx.getText();
        sb.append(var);
        parameters.add(var);
        return null;
    }

    @Override
    public Void visitKeyword(final GremlinParser.KeywordContext ctx) {
        final String keyword = ctx.getText();

        // translate differently based on the context of the keyword's parent.
        if (ctx.getParent() instanceof GremlinParser.MapKeyContext || ctx.getParent() instanceof GremlinParser.ConfigurationContext) {
            // if the keyword is a key in a map, then it's a string literal essentially
            sb.append(keyword);
        } else {
            // in all other cases it's used more like "new Class()"
            sb.append(keyword).append(" ");
        }

        return null;
    }

    @Override
    public Void visitTerminal(final TerminalNode node) {
        // skip EOF node
        if (null == node || node.getSymbol().getType() == -1) return null;

        final String terminal = node.getSymbol().getText();
        switch (terminal) {
            case "(":
                appendStepOpen();
                break;
            case ")":
                appendStepClose();
                break;
            case ",":
                appendArgumentSeparator();
                break;
            case ".":
                appendStepSeparator();
                break;
            case "new":
                // seems we still sometimes interpret this as a TerminalNode like when newing up a class
                // which is optional syntax and will go away in the future.
                sb.append("new");
                if (!(node.getParent() instanceof GremlinParser.MapKeyContext))
                    sb.append(" "); // includes a space for when not use in context of a Map entry key...one off
                break;
            default:
                sb.append(processGremlinSymbol(terminal));
        }
        return null;
    }

    @Override
    public Void visitTraversalTShort(final GremlinParser.TraversalTShortContext ctx) {
        appendExplicitNaming(ctx.getText(), T.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalTLong(final GremlinParser.TraversalTLongContext ctx) {
        appendExplicitNaming(ctx.getText(), T.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalDirectionShort(final GremlinParser.TraversalDirectionShortContext ctx) {
        appendExplicitNaming(ctx.getText(), Direction.class.getSimpleName());
        return null;
    }

    @Override
    public Void visitTraversalDirectionLong(final GremlinParser.TraversalDirectionLongContext ctx) {
        appendExplicitNaming(ctx.getText(), Direction.class.getSimpleName());
        return null;
    }

    protected void appendExplicitNaming(final String txt, final String prefix) {
        if (!txt.startsWith(prefix + ".")) {
            sb.append(processGremlinSymbol(prefix)).append(".");
            sb.append(processGremlinSymbol(txt));
        } else {
            final String[] split = txt.split("\\.");
            sb.append(processGremlinSymbol(split[0])).append(".");
            sb.append(processGremlinSymbol(split[1]));
        }
    }

    protected void appendAnonymousSpawn() {
        sb.append("__.");
    }

    protected String getCardinalityFunctionClass() {
        return Cardinality.class.getSimpleName();
    }
}

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/language/translator/TranslateVisitor.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { ErrorNode, TerminalNode } from 'antlr4ng';
import { GremlinLexer } from '../grammar/GremlinLexer.js';

/**
 * Base visitor for translating Gremlin queries. Mirrors the Java TranslateVisitor.
 * Makes no changes to input except:
 * - Normalizes whitespace
 * - Normalizes numeric suffixes to lower case
 * - Makes anonymous traversals explicit with double underscore
 * - Makes enums explicit with their proper name
 */
export default class TranslateVisitor {

    protected readonly graphTraversalSourceName: string;
    protected readonly sb: string[] = [];
    protected readonly parameters: Set<string> = new Set();

    constructor(graphTraversalSourceName: string = 'g') {
        this.graphTraversalSourceName = graphTraversalSourceName;
    }

    getTranslated(): string {
        return this.sb.join('');
    }

    getParameters(): ReadonlySet<string> {
        return this.parameters;
    }

    protected processGremlinSymbol(step: string): string {
        return step;
    }

    protected appendArgumentSeparator(): void {
        this.sb.push(', ');
    }

    protected appendStepSeparator(): void {
        this.sb.push('.');
    }

    protected appendStepOpen(): void {
        this.sb.push('(');
    }

    protected appendStepClose(): void {
        this.sb.push(')');
    }

    protected appendAnonymousSpawn(): void {
        this.sb.push('__.');
    }

    visit(tree: any): void {
        if (tree == null) return;

        // antlr-ng does not generate accept() overrides per context class —
        // ParserRuleContext.accept() always calls visitChildren, never dispatching to
        // the specific visitXxx method. We perform the dispatch ourselves here by
        // mapping the context class name (e.g. StringLiteralContext) to the
        // corresponding visitor method (e.g. visitStringLiteral).
        const className: string = tree?.constructor?.name ?? '';
        if (className.endsWith('Context')) {
            const methodName = 'visit' + className.slice(0, -'Context'.length);
            const method = (this as any)[methodName];
            if (typeof method === 'function') {
                method.call(this, tree);
                return;
            }
        }

        // No specific visitor method found — fall through to accept() which calls
        // visitChildren for rule contexts, visitTerminal for terminals.
        (tree as any).accept(this);
    }

    visitChildren(node: any): void {
        const n = node.getChildCount();
        for (let i = 0; i < n; i++) {
            this.visit(node.getChild(i));
        }
    }

    protected getCardinalityFunctionClass(): string {
        return 'Cardinality';
    }

    protected static removeFirstAndLastCharacters(text: string): string {
        return text != null && text.length > 0 ? text.substring(1, text.length - 1) : '';
    }

    protected handleStringLiteralText(text: string): void {
        this.sb.push('"');
        this.sb.push(text);
        this.sb.push('"');
    }

    protected appendExplicitNaming(txt: string, prefix: string): void {
        if (!txt.startsWith(prefix + '.')) {
            this.sb.push(this.processGremlinSymbol(prefix));
            this.sb.push('.');
            this.sb.push(this.processGremlinSymbol(txt));
        } else {
            const parts = txt.split('.');
            this.sb.push(this.processGremlinSymbol(parts[0]));
            this.sb.push('.');
            this.sb.push(this.processGremlinSymbol(parts[1]));
        }
    }

    visitTerminal(node: TerminalNode): void {
        // skip EOF node
        if (node == null || node.getSymbol().type === -1) return;

        // TRAVERSAL_ROOT is the traversal source token (e.g. 'g') — always output as-is,
        // never subject to processGremlinSymbol (which may capitalize it in some dialects).
        if (node.getSymbol().type === GremlinLexer.TRAVERSAL_ROOT) {
            this.sb.push(this.graphTraversalSourceName);
            return;
        }

        const terminal = node.getSymbol().text ?? '';
        switch (terminal) {
            case '(':
                this.appendStepOpen();
                break;
            case ')':
                this.appendStepClose();
                break;
            case ',':
                this.appendArgumentSeparator();
                break;
            case '.':
                this.appendStepSeparator();
                break;
            case 'new':
                this.sb.push('new');
                if (!(node.parent?.constructor?.name === 'MapKeyContext')) {
                    this.sb.push(' ');
                }
                break;
            default:
                this.sb.push(this.processGremlinSymbol(terminal));
        }
    }

    visitErrorNode(_node: ErrorNode): void {
        // no-op
    }

    visitTraversalSource(ctx: any): void {
        this.sb.push(this.graphTraversalSourceName);
        // child counts more than 1 means there is a step separator and a traversal source method
        if (ctx.getChildCount() > 1) {
            if (ctx.getChild(0).getChildCount() > 1) {
                this.appendStepSeparator();
                this.visitTraversalSourceSelfMethod(ctx.getChild(0).getChild(2));
            }
            this.appendStepSeparator();
            this.visitTraversalSourceSelfMethod(ctx.getChild(2));
        }
    }

    visitTraversalSourceSelfMethod(ctx: any): void {
        this.visitChildren(ctx);
    }

    visitNestedTraversal(ctx: any): void {
        if (ctx.ANON_TRAVERSAL_ROOT() == null) {
            this.appendAnonymousSpawn();
        }
        this.visitChildren(ctx);
    }

    visitTraversalScope(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Scope');
    }

    visitTraversalT(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'T');
    }

    visitTraversalTShort(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'T');
    }

    visitTraversalTLong(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'T');
    }

    visitTraversalMerge(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Merge');
    }

    visitTraversalOrder(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Order');
    }

    visitTraversalBarrier(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Barrier');
    }

    visitTraversalDirection(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Direction');
    }

    visitTraversalDirectionShort(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Direction');
    }

    visitTraversalDirectionLong(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Direction');
    }

    visitTraversalCardinality(ctx: any): void {
        if (ctx.LPAREN() != null && ctx.RPAREN() != null) {
            const idx = ctx.K_CARDINALITY() != null ? 2 : 0;
            const txt = ctx.getChild(idx).getText();
            this.appendExplicitNaming(txt, this.getCardinalityFunctionClass());
            this.appendStepOpen();
            this.visit(ctx.genericLiteral());
            this.appendStepClose();
        } else {
            this.appendExplicitNaming(ctx.getText(), 'Cardinality');
        }
    }

    visitTraversalColumn(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Column');
    }

    visitTraversalPop(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Pop');
    }

    visitTraversalOperator(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Operator');
    }

    visitTraversalPick(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'Pick');
    }

    visitTraversalDT(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'DT');
    }

    visitTraversalGType(ctx: any): void {
        this.appendExplicitNaming(ctx.getText(), 'GType');
    }

    visitTraversalPredicate(ctx: any): void {
        switch (ctx.getChildCount()) {
            case 1:
                this.visit(ctx.getChild(0));
                break;
            case 5:
                // handle negate of P
                this.visit(ctx.getChild(0));
                this.sb.push('.');
                this.sb.push(this.processGremlinSymbol('negate'));
                this.sb.push('()');
                break;
            case 6:
                // handle and/or predicates
                this.visit(ctx.getChild(0));
                this.sb.push('.');
                this.sb.push(this.processGremlinSymbol(ctx.getChild(2).getText()));
                this.sb.push('(');
                this.visit(ctx.getChild(4));
                this.sb.push(')');
                break;
        }
    }

    protected visitP(ctx: any, pClass: string, methodName: string): void {
        this.sb.push(pClass);
        this.appendStepSeparator();
        this.sb.push(this.processGremlinSymbol(methodName));
        this.appendStepOpen();

        const children: any[] = [];
        for (let i = 0; i < ctx.getChildCount(); i++) {
            const child = ctx.getChild(i);
            const name = child.constructor?.name ?? '';
            if (name === 'GenericArgumentContext' ||
                name === 'GenericArgumentVarargsContext' ||
                name === 'StringArgumentContext' ||
                name === 'StringLiteralContext' ||
                name === 'TraversalGTypeContext' ||
                name === 'TraversalPredicateContext') {
                children.push(child);
            }
        }

        for (let ix = 0; ix < children.length; ix++) {
            this.visit(children[ix]);
            if (ix < children.length - 1) {
                this.appendArgumentSeparator();
            }
        }

        this.appendStepClose();
    }

    visitTraversalPredicate_eq(ctx: any): void { this.visitP(ctx, 'P', 'eq'); }
    visitTraversalPredicate_neq(ctx: any): void { this.visitP(ctx, 'P', 'neq'); }
    visitTraversalPredicate_lt(ctx: any): void { this.visitP(ctx, 'P', 'lt'); }
    visitTraversalPredicate_lte(ctx: any): void { this.visitP(ctx, 'P', 'lte'); }
    visitTraversalPredicate_gt(ctx: any): void { this.visitP(ctx, 'P', 'gt'); }
    visitTraversalPredicate_gte(ctx: any): void { this.visitP(ctx, 'P', 'gte'); }
    visitTraversalPredicate_inside(ctx: any): void { this.visitP(ctx, 'P', 'inside'); }
    visitTraversalPredicate_outside(ctx: any): void { this.visitP(ctx, 'P', 'outside'); }
    visitTraversalPredicate_between(ctx: any): void { this.visitP(ctx, 'P', 'between'); }
    visitTraversalPredicate_within(ctx: any): void { this.visitP(ctx, 'P', 'within'); }
    visitTraversalPredicate_without(ctx: any): void { this.visitP(ctx, 'P', 'without'); }
    visitTraversalPredicate_typeOf(ctx: any): void { this.visitP(ctx, 'P', 'typeOf'); }
    visitTraversalPredicate_not(ctx: any): void { this.visitP(ctx, 'P', 'not'); }
    visitTraversalPredicate_containing(ctx: any): void { this.visitP(ctx, 'TextP', 'containing'); }
    visitTraversalPredicate_notContaining(ctx: any): void { this.visitP(ctx, 'TextP', 'notContaining'); }
    visitTraversalPredicate_startingWith(ctx: any): void { this.visitP(ctx, 'TextP', 'startingWith'); }
    visitTraversalPredicate_notStartingWith(ctx: any): void { this.visitP(ctx, 'TextP', 'notStartingWith'); }
    visitTraversalPredicate_endingWith(ctx: any): void { this.visitP(ctx, 'TextP', 'endingWith'); }
    visitTraversalPredicate_notEndingWith(ctx: any): void { this.visitP(ctx, 'TextP', 'notEndingWith'); }
    visitTraversalPredicate_regex(ctx: any): void { this.visitP(ctx, 'TextP', 'regex'); }
    visitTraversalPredicate_notRegex(ctx: any): void { this.visitP(ctx, 'TextP', 'notRegex'); }

    visitBooleanArgument(ctx: any): void {
        if (ctx.booleanLiteral() != null) {
            this.visitBooleanLiteral(ctx.booleanLiteral());
        } else {
            this.visitVariable(ctx.variable());
        }
    }

    visitGenericArgument(ctx: any): void {
        if (ctx.genericLiteral() != null) {
            this.visitGenericLiteral(ctx.genericLiteral());
        } else {
            this.visitVariable(ctx.variable());
        }
    }

    visitGenericLiteral(ctx: any): void {
        this.visitChildren(ctx);
    }

    visitIntegerLiteral(ctx: any): void {
        this.sb.push(ctx.getText().toLowerCase());
    }

    visitFloatLiteral(ctx: any): void {
        if (ctx.infLiteral() != null) { this.visit(ctx.infLiteral()); return; }
        if (ctx.nanLiteral() != null) { this.visit(ctx.nanLiteral()); return; }
        this.sb.push(ctx.getText().toLowerCase());
    }

    visitBooleanLiteral(ctx: any): void {
        this.sb.push(ctx.getText());
    }

    visitNullLiteral(ctx: any): void {
        this.sb.push(ctx.getText());
    }

    visitNanLiteral(ctx: any): void {
        this.sb.push(ctx.getText());
    }

    visitInfLiteral(ctx: any): void {
        this.sb.push(ctx.getText());
    }

    visitUuidLiteral(ctx: any): void {
        this.sb.push(ctx.getText());
    }

    visitStringLiteral(ctx: any): void {
        const text = TranslateVisitor.removeFirstAndLastCharacters(ctx.getText());
        this.handleStringLiteralText(text);
    }

    visitStringNullableLiteral(ctx: any): void {
        if (ctx.getText() === 'null') {
            this.sb.push('null');
        } else {
            const text = TranslateVisitor.removeFirstAndLastCharacters(ctx.getText());
            this.handleStringLiteralText(text);
        }
    }

    visitNakedKey(ctx: any): void {
        this.handleStringLiteralText(ctx.getText());
    }

    visitVariable(ctx: any): void {
        const varName: string = ctx.getText();
        this.sb.push(varName);
        this.parameters.add(varName);
    }

    visitKeyword(ctx: any): void {
        const keyword: string = ctx.getText();
        if (ctx.parent?.constructor?.name === 'MapKeyContext' ||
            ctx.parent?.constructor?.name === 'ConfigurationContext') {
            this.handleStringLiteralText(keyword);
        } else {
            this.sb.push(keyword);
            this.sb.push(' ');
        }
    }

    visitMapKey(ctx: any): void {
        const keyIndex = (ctx.LPAREN() != null && ctx.RPAREN() != null) ? 1 : 0;
        this.visit(ctx.getChild(keyIndex));
    }

    visitMapEntry(ctx: any): void {
        this.visitChildren(ctx);
    }
}
